wandb/sdk_py27/wandb_artifacts.py

```diff
@@ -1175,61 +1153,122 @@ class WBArtifactHandler(StorageHandler):
     def __init__(self, scheme=None):
         self._scheme = scheme or "wandb-artifact"
         self._cache = get_artifacts_cache()
+        self._client = None
 
     @property
     def scheme(self):
+        """overrides parent scheme
+        :return: The scheme to which this handler applies.
+        :rtype: str
+        """
         return self._scheme
 
     @staticmethod
     def parse_path(path):
+        """helper method to parse a URL into it's component parts."""
+
         url = urlparse(path)
         return url.netloc, (url.path if url.path[0] != "/" else url.path[1:])
 
+    @property
+    def client(self):
+        if self._client is None:
+            self._client = PublicApi()
+        return self._client
+
     def load_path(self, artifact, manifest_entry, local=False):
+        """
+        Loads the file within the specified artifact given its
+        corresponding entry. In this case, the referenced artifact is downloaded
+        and a new symlink is created and returned to the caller.
+
+        :param manifest_entry: The index entry to load
+        :type manifest_entry: ArtifactManifestEntry
+        :return: A path to the file represented by `index_entry`
+        :rtype: os.PathLike
+        """
+
+        # Standard shortcircut if the asset is already downloaded
         path, hit = self._cache.check_etag_obj_path(
             manifest_entry.digest, manifest_entry.size
         )
         if hit:
             return path
 
+        # Parse the reference path and download the artifact if needed
         artifact_id, artifact_file_path = WBArtifactHandler.parse_path(
             manifest_entry.ref
         )
-        artifact = PublicApi().artifact_from_id(util.hex_to_b64_id(artifact_id))
-        artifact_path = artifact.download()
 
-        link_target_path = os.path.join(artifact_path, artifact_file_path)
+        dep_artifact = PublicArtifact.from_id(
+            util.hex_to_b64_id(artifact_id), self.client
+        )
+        link_target_path = dep_artifact.get_path(artifact_file_path).download()
         link_creation_path = os.path.join(
-            self._cache._cache_dir, "tmp", link_target_path
+            # This tmp directory is created in order to have a place to put the symlink.
+            # Since this is a threaded operation, I was getting collisions and needed to create
+            # a unique directory which would not collide even with the same reference target.
+            self._cache._cache_dir,
```

Comment:
Would you mind explaining the collisions you were seeing?

Also, shouldn't the symlink be targeting `path`? We'll always have a cache miss otherwise, right?

Reply:
I actually removed all the symlinking stuff per Shawn's request, so this is now gone. The TLDR is that at the time of loading, you only know the target asset you want to link to, so if there are multiple references to the same asset (for example multiple references to the same Image or classes file), then the symlink generated location would collide. Using the thread ID as a path component removed this collision.